### PR TITLE
HamiltonianGate

### PR DESCRIPTION
**Pull request by dongreenberg**
Introducing HamiltonianGate, and tests. Tests pass. Also allows complex resolution of ParameterExpression.

### Summary

HamiltonianGate takes a hermitian matrix or Operator and time parameter (I wanted to name this `t` but pylint didn't let me), and resolves to a Unitary gate equal to e^iHt. This allows such evolutions to be added to a circuit and the time parameter bound later.

Context: I believe this will be a major performance improvement in Aqua simulation, avoiding constructing large evolution circuits for ansatze like UCCSD.

### CODE REVIEW

**Comment by ajavadia** [comment]
> I wanted to name this t but pylint didn't let me

You can edit `.pylintrc` and insert `t` as a `good-name`

**Comment by dongreenberg** [comment]
@ajavadia should I?

File: qiskit/circuit/parameterexpression.py [hunk 1]
```diff
@@ -236,6 +236,12 @@ def __float__(self):
                             'cannot be cast to a float.'.format(self.parameters))
         return float(self._symbol_expr)
 
+    def __complex__(self):
+        if self.parameters:
+            raise TypeError('ParameterExpression with unbound parameters ({}) '
+                            'cannot be cast to a float.'.format(self.parameters))
```
Comment (Cryoris) [inline qiskit/circuit/parameterexpression.py:242]: Should this be complex?
```suggestion
                            'cannot be cast to a complex.'.format(self.parameters))
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:5]:
```suggestion
# (C) Copyright IBM 2017, 2020.
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            label (str): unitary name for backend [Default: None].
+
+        Raises:
+            ExtensionError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, 'to_matrix'):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:53]: Would it be safer to check for types instead of attributes here? The `to_matrix` is probably safe both ways but `to_operator` might also be the attribute of another class than `BaseOperator`.

File: qiskit/extensions/hamiltonian_gate.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        return UnitaryGate(self.to_matrix())
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:117]: This shouldn't return anything afaik, but just set `self.definition` instead
```suggestion
    def _define(self):
        """Calculate a subcircuit that implements this unitary."""
        self.definition = UnitaryGate(self.to_matrix()).definition
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        Returns:
+            UnitaryGate: controlled version of gate.
+
+        Raises:
+            QiskitError: invalid ctrl_state
+        """
+        return self._define().control(num_ctrl_qubits=num_ctrl_qubits,
+                                      label=label,
+                                      ctrl_state=ctrl_state)
+
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:137]: If you implement `_define` as is done with the other gates you should be able to use `control` from the base class 🙂

File: qiskit/extensions/hamiltonian_gate.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    def qasm(self):
+        """ The qasm for the UnitaryGate defining this HamiltonianGate
+        """
+        return self._define().qasm()
+
+
+def hamiltonian(self, matrix, time, qubits, label=None):
+    """Apply u2 to q."""
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:145]: This docstring must be updated

File: qiskit/extensions/hamiltonian_gate.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    if isinstance(qubits, QuantumRegister):
+        qubits = qubits[:]
```
Comment (Cryoris) [inline qiskit/extensions/hamiltonian_gate.py:147]: I think passing a quantum register is already supported if it is passed as
```
self.append(H, [quantum_register], [])
```
However this probably doesn't work if we apply a hamiltionian on just one qubit, such like 
```python
circuit.hamiltionian(H, time=0.5, qubits=0)
```
I think the solution here would be
```python
if not isinstance(qubits, list):
    qubits = [qubits]

return self.append(HamiltonianGate(data=matrix, time=time, label=label), qubits, [])
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Comment (Cryoris) [inline test/python/circuit/test_hamiltonian_gate.py:5]:
```suggestion
# (C) Copyright IBM 2017, 2020.
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+from qiskit.transpiler.passes import CXCancellation
+
+
+class TestHamiltonianGate(QiskitTestCase):
+    """Tests for the HamiltonianGate class."""
+
+    def setUp(self):
+        """Setup."""
+        pass
+
```
Comment (Cryoris) [inline test/python/circuit/test_hamiltonian_gate.py:41]: Not required if doesn't do anything
```suggestion
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    def test_set_matrix(self):
+        """Test instantiation"""
+        try:
+            HamiltonianGate([[0, 1], [1, 0]], 1)
+        # pylint: disable=broad-except
+        except Exception as err:
+            self.fail('unexpected exception in init of HamiltonianGate: {0}'.format(err))
```
Comment (Cryoris) [inline test/python/circuit/test_hamiltonian_gate.py:48]: A more elegant way to test this could be
```suggestion
        hamiltonian = HamiltonianGate([[0, 1], [1, 0]], 1)
        self.assertEqual(hamiltonian.num_qubits, 1)
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+    def test_set_matrix_raises(self):
+        """test non-unitary"""
+        try:
+            HamiltonianGate([[1, 0], [1, 1]], 1)
+        # pylint: disable=broad-except
+        except Exception:
+            pass
+        else:
+            self.fail('setting Unitary with non-hermitian matrix did not raise')
```
Comment (Cryoris) [inline test/python/circuit/test_hamiltonian_gate.py:58]: You could use `assertRaises` here
```suggestion
        with self.assertRaises(ExtensionError):
            HamiltonianGate([[1, 0], [1, 1]], 1)
```

File: test/python/quantum_info/states/test_measures.py [hunk 12]
```diff
@@ -187,7 +187,7 @@ def test_entropy_density_matrix(self):
         for rho in rhos:
             self.assertAlmostEqual(entropy(rho), 1)
             self.assertAlmostEqual(entropy(rho, base=2), 1)
-            self.assertAlmostEqual(entropy(rho, base=np.e), -np.log(0.5))
+            self.assertAlmostEqual(entropy(rho, base=np.e), -1 * np.log(0.5))
```
Comment (Cryoris) [inline test/python/quantum_info/states/test_measures.py:190]: Why was this added?

File: qiskit/extensions/hamiltonian_gate.py (hunk 3 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:53]: I was trying to just replicate the API for UnitaryGate. Do you think it doesn't make sense here?

File: qiskit/extensions/hamiltonian_gate.py (hunk 4 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:117]: Great point, thanks!

File: test/python/quantum_info/states/test_measures.py (hunk 12 above)
Comment (dongreenberg) [reply test/python/quantum_info/states/test_measures.py:190]: Lint error about applying negative to type which doesn't support negative.

File: qiskit/extensions/hamiltonian_gate.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        """Return matrix for the unitary."""
+        unbound_t = False
+        if isinstance(self.params[1], ParameterExpression):
+            unbound_t = len(self.params[1].parameters) > 0
+        if isinstance(self.params[1], Number) or not unbound_t:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(1j * self.params[0] * self.params[1])
+        else:
+            raise NotImplementedError("Unable to generate Unitary matrix for "
+                                      "unbound t parameter {}".format(self.params[1]))
```
Comment (kdk) [inline qiskit/extensions/hamiltonian_gate.py:93]:
```suggestion
        try:
            # pylint: disable=no-member
            return scipy.linalg.expm(1j * self.params[0] * float(self.params[1]))
         except TypeError:
            raise TypeError("Unable to generate Unitary matrix for "
                            "unbound t parameter {}".format(self.params[1]))
```
Calling `float` on the `ParameterExpression` will raise a `TypeError` if it is unbound, and will give you  float you can multiple by `1j` (so you won't need the parameter changes).

File: qiskit/extensions/hamiltonian_gate.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        self.definition = UnitaryGate(self.to_matrix()).definition
```
Comment (kdk) [inline qiskit/extensions/hamiltonian_gate.py:113]: `.definition` needs to be a list of `(Instruction, qargs, cargs)` tuples. See e.g. the definition for `HGate` in terms of `U2Gate`

File: qiskit/extensions/hamiltonian_gate.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+        # Store instruction params
+        super().__init__('hamiltonian', num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=True)
```
Comment (kdk) [inline qiskit/extensions/hamiltonian_gate.py:75]:
```suggestion
        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=False)
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+            def default(self, obj):
+                if isinstance(obj, np.ndarray):
+                    return obj.tolist()
+                if isinstance(obj, complex):
+                    return (obj.real, obj.imag)
+                return json.JSONEncoder.default(self, obj)
+
+        # check json serialization
+        self.assertTrue(isinstance(json.dumps(qobj_dict, cls=NumpyEncoder),
+                                   str))
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:181]: Can you add a test that this Decomposes to the right UnitaryGate?

File: test/python/circuit/test_hamiltonian_gate.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        cr = ClassicalRegister(2)
+        qc = QuantumCircuit(qr, cr)
+        matrix = Operator.from_label('XY')
+        qc.x(qr[0])
+        theta = Parameter('theta')
+        uni2q = HamiltonianGate(matrix, theta)
+        qc.append(uni2q, [qr[0], qr[1]])
+        passman = PassManager()
+        passman.append(CXCancellation())
+        qc2 = transpile(qc, pass_manager=passman)
```
[the remark this answers is not in the document]
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:107]: Yes, to make sure we can transpile before the bind.

File: qiskit/extensions/hamiltonian_gate.py (hunk 15 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:75]: Good catch

File: test/python/circuit/test_hamiltonian_gate.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+        np.testing.assert_array_almost_equal(ham.adjoint().to_matrix(),
+                                             np.transpose(np.conj(ham.to_matrix())))
+
+
+class TestHamiltonianCircuit(QiskitTestCase):
+    """Hamiltonian gate circuit tests."""
+
+    def test_1q_hamiltonian(self):
+        """test 1 qubit hamiltonian"""
+        qr = QuantumRegister(1)
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:74]:
```suggestion
        qr = QuantumRegister(1, 'q0')
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        cr = ClassicalRegister(1)
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:75]:
```suggestion
        cr = ClassicalRegister(1, 'c0')
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                   "include \"qelib1.inc\";\n" \
+                   "qreg q0[1];\n" \
+                   "creg c0[1];\n" \
+                   "x q0[0];\n" \
+                   "hamiltonian([[0.+0.j 0.+0.j]\n" \
+                   " [0.+0.j 0.+0.j]],1) q0[0];\n"
+        self.assertEqual(qc.qasm(), ref_qasm)
+
+        # test of text drawer
+        self.log.info(qc)
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:95]: You can assert on the drawn circuit including a HamiltonianGate (see examples in `test/python/visualization/test_circuit_text_drawer.py`).

Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:95]: Good catch - this prints pretty wonky (below) because it prints with the Hermitian matrix. Do you know how to get it to print with just the name (and maybe parameter name?) like Unitary does?
```
      ┌───┐┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
q1_0: ┤ X ├┤0                                                                                                                                                            ├
      └───┘│  Hamiltonian([[0.+0.j 0.+0.j 0.+0.j 0.-1.j]
 [0.+0.j 0.+0.j 0.+1.j 0.+0.j]
 [0.+0.j 0.-1.j 0.+0.j 0.+0.j]
 [0.+1.j 0.+0.j 0.+0.j 0.+0.j]],1.57079632679490) │
q1_1: ─────┤1                                                                                                                                                            ├
           └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
c1_0: ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
                                                                                                                                                                          
c1_1: ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
```

File: test/python/circuit/test_hamiltonian_gate.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+        qc = qc.bind_parameters({theta: 1})
+
+        # test of qasm output
+        self.log.info(qc.qasm())
+        ref_qasm = "OPENQASM 2.0;\n" \
+                   "include \"qelib1.inc\";\n" \
+                   "qreg q0[1];\n" \
+                   "creg c0[1];\n" \
+                   "x q0[0];\n" \
+                   "hamiltonian([[0.+0.j 0.+0.j]\n" \
```
Comment (1ucian0) [inline test/python/circuit/test_hamiltonian_gate.py:90]: Strictly speaking, this is not OpenQASM 2.0. I'm not sure if this can be solved like in UnitaryGate, tho...

File: test/python/circuit/test_hamiltonian_gate.py (hunk 20 above)
Comment (1ucian0) [reply test/python/circuit/test_hamiltonian_gate.py:95]: By adding a label (`qc.append(HamiltonianGate(matrix, theta, label='Hamiltonian'), [qr[0]])` you will get the following result:
```
      ┌───┐┌─────────────┐
q0_0: ┤ X ├┤ Hamiltonian ├
      └───┘└─────────────┘
c0_0: ════════════════════
```
Not sure if that helps.

File: test/python/circuit/test_hamiltonian_gate.py [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+            """Class for encoding json str with complex and np arrays."""
+            def default(self, obj):
+                if isinstance(obj, np.ndarray):
+                    return obj.tolist()
+                if isinstance(obj, complex):
+                    return (obj.real, obj.imag)
+                return json.JSONEncoder.default(self, obj)
+
+        # check json serialization
+        self.assertTrue(isinstance(json.dumps(qobj_dict, cls=NumpyEncoder),
```
Comment (1ucian0) [inline test/python/circuit/test_hamiltonian_gate.py:189]: The encoder is defined as part of the test and I dont see the point of testing it. Either the encoder should be part of the code base or the test is testing the test...

Comment (mtreinish) [reply test/python/circuit/test_hamiltonian_gate.py:189]: This is basically what the ibmq provider uses internally now that qobj is marshmallow free: https://github.com/Qiskit/qiskit-ibmq-provider/blob/2c4d04dc91ddd495643a69c82dd92434bfc21603/qiskit/providers/ibmq/utils/json_encoder.py so I think it's fine to test this if we need to. It's something we do in the qobj tests for example. But I agree in this case testing that the qobj object post json serialization doesn't provide too much value. We're not actually testing anything outside of what is already covered by qobj testing and ibmq provider testing. But in those cases we test that the object is preserved through serialization (by reloading it) just asserting it's a string doesn't seem to useful. That being said I'm not sure if it's really a blocker here, its something we can always cleanup in a follow up pr.

File: qiskit/extensions/hamiltonian_gate.py (hunk 3 above)
Comment (chriseclectic) [reply qiskit/extensions/hamiltonian_gate.py:50]: To operator lets you initialize with a quantum channel object in the case where that channel is really a unitary channel. This might be a not-needed edge case for this gate.

File: qiskit/extensions/hamiltonian_gate.py [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+            return False
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=False)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def to_matrix(self):
+        """Return matrix for the unitary."""
+        try:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(1j * self.params[0] * float(self.params[1]))
```
Comment (chriseclectic) [inline qiskit/extensions/hamiltonian_gate.py:83]: Time evolution is usually defined with a minus sign
```suggestion
            return scipy.linalg.expm(-1j * self.params[0] * float(self.params[1]))
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to numpy array in case not already an array
+        data = numpy.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
```
Comment (chriseclectic) [inline qiskit/extensions/hamiltonian_gate.py:58]: We should probably throw an exception if time is complex or imaginary too

File: qiskit/extensions/hamiltonian_gate.py [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+            raise TypeError("Unable to generate Unitary matrix for "
+                            "unbound t parameter {}".format(self.params[1]))
+
+    def inverse(self):
+        """Return the adjoint of the unitary."""
+        return self.adjoint()
+
+    def conjugate(self):
+        """Return the conjugate of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), -self.params[1])
```
Comment (chriseclectic) [inline qiskit/extensions/hamiltonian_gate.py:94]: Might be clearer here to conjugate the hamiltonian here instead of transpose (even though they are equivalent for hermitian matrices)

File: qiskit/extensions/hamiltonian_gate.py [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+    """Class for representing evolution by a Hermitian Hamiltonian operator as a gate. This gate
+    resolves to a UnitaryGate, which can be decomposed into basis gates if it is 2 qubits or
+    less, or simulated directly in Aer for more qubits. """
+
+    def __init__(self, data, time, label=None):
+        """Create a gate from a hamiltonian operator and evolution time parameter t
+
+        Args:
+            data (matrix or Operator): unitary operator.
+            time (float or complex): time evolution parameter.
```
[the remark this answers is not in the document]
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:40]: This is to support imaginary time evolution.

File: qiskit/extensions/hamiltonian_gate.py (hunk 25 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:94]: Good call, done

File: qiskit/extensions/hamiltonian_gate.py (hunk 3 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:50]: Ok, removing.

File: qiskit/extensions/hamiltonian_gate.py (hunk 23 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:83]: Done

File: qiskit/extensions/hamiltonian_gate.py (hunk 3 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:50]: Actually - I need something in here to be able to pass in Operators anyway, so is it a problem to leave it as is to handle the broader set of BaseOperators, or would it be better to change it to an isinstance(Operator)?

File: test/python/circuit/test_hamiltonian_gate.py (hunk 21 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:90]: I don't think it can. Should I just take out the ability to generate QASM and throw an error if the user tries?

File: test/python/circuit/test_hamiltonian_gate.py (hunk 20 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:95]: Ok - so should we say this is WAI? Or should I add a label in the constructor if none is provided? Similar to: `self._qasm_name = self.label if self.label else "unitary" + str(id(self))` in the Unitary qasm generator?

Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:95]: Or only add the label if none is given in the `qc.hamiltonian()` monkeypatch?

Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:95]: Or just change the default label in the signature from None to "Hamiltonian"?

**Comment by nonhermitian** [comment]
This can quickly lead to trouble if the user does not use normalized units for their Hamiltonian.  For example:
```python
A = array([[1000,  0],
           [0, 1000]])
```
fails because of the `expm`

**Comment by dongreenberg** [comment]
> This can quickly lead to trouble if the user does not use normalized units for their Hamiltonian. For example:
> 
> ```python
> A = array([[1000,  0],
>            [0, 1000]])
> ```
> 
> fails because of the `expm`

I tried this and it seems to work ok. What's the failure you're thinking of?

**Comment by nonhermitian** [comment]
Namely that this

 https://github.com/Qiskit/qiskit-terra/blob/00a53690e5c06d66fb0bc42a5059a999add98179/qiskit/extensions/hamiltonian_gate.py#L82

will fail.

File: test/python/circuit/test_hamiltonian_gate.py (hunk 21 above)
Comment (1ucian0) [reply test/python/circuit/test_hamiltonian_gate.py:94]: sounds good to me.. maybe @ajavadia or @kdk have an opinion here.

**Comment by dongreenberg** [comment]
I see this:
```
>>> A = np.asarray([[1000,  0],[0, 1000]])
>>> scipy.linalg.expm(-1j * A)
Out[6]: 
array([[0.56237908-0.82687954j, 0.        +0.j        ],
       [0.        +0.j        , 0.56237908-0.82687954j]])
```
Am I missing something?

File: test/python/circuit/test_hamiltonian_gate.py (hunk 20 above)
Comment (kdk) [reply test/python/circuit/test_hamiltonian_gate.py:99]: I think this is WAI, but there is a bug here: https://github.com/Qiskit/qiskit-terra/blob/80ca284048e2e69e1e9eb8b92112dccefeeb2f64/qiskit/visualization/text.py#L706

We omit array parameters from drawing only if all parameters are ndarrays. We can fix that here or in its own issue.

File: test/python/circuit/test_hamiltonian_gate.py (hunk 21 above)
Comment (kdk) [reply test/python/circuit/test_hamiltonian_gate.py:94]: Okay with raising an error on `qasm()` (and a follow up issue to try to find a better way). For `UnitaryGate`, we synthesize on call to `qasm` (because QASM doesn't support array-like parameters). Maybe a similar hack could work here.

File: test/python/circuit/test_hamiltonian_gate.py (hunk 20 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:99]: Luciano actually just fixed that and pushed to this PR, so we're all good.

File: test/python/circuit/test_hamiltonian_gate.py (hunk 21 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:94]: Ok, done.

File: test/python/circuit/test_hamiltonian_gate.py [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+        nodes = dag.two_qubit_ops()
+        self.assertEqual(len(nodes), 1)
+        dnode = nodes[0]
+        self.assertIsInstance(dnode.op, HamiltonianGate)
+        for qubit in dnode.qargs:
+            self.assertIn(qubit.index, [0, 1])
+        # Equality based on Pauli exponential identity
+        np.testing.assert_array_almost_equal(dnode.op.to_matrix(), 1j*matrix.data)
+        qc3 = dag_to_circuit(dag)
+        self.assertEqual(qc2, qc3)
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:136]: Not blocking, but the tests would be easier to follow if they were each a bit narrower in scope. Ideally, covering a single property or use-case per test.

File: test/python/circuit/test_hamiltonian_gate.py [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+        """test qobj output with hamiltonian"""
+        qr = QuantumRegister(4)
+        qc = QuantumCircuit(qr)
+        qc.rx(np.pi / 4, qr[0])
+        matrix = Operator.from_label('XIZ')
+        theta = Parameter('theta')
+        uni = HamiltonianGate(matrix, theta, label='XIZ')
+        qc.append(uni, [qr[0], qr[1], qr[3]])
+        qc.cx(qr[3], qr[2])
+        qc = qc.bind_parameters({theta: -np.pi / 2})
```
Comment (kdk) [inline test/python/circuit/test_hamiltonian_gate.py:170]:
```suggestion
        qc = qc.bind_parameters({theta: np.pi / 2})
```

File: qiskit/extensions/hamiltonian_gate.py [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+
+def hamiltonian(self, operator, time, qubits, label=None):
+    """Apply hamiltonian evolution to to qubits."""
+    if not isinstance(qubits, list):
+        qubits = [qubits]
+
+    return self.append(HamiltonianGate(data=operator, time=time, label=label), qubits, [])
+
+
+QuantumCircuit.hamiltonian = hamiltonian
```
Comment (kdk) [inline qiskit/extensions/hamiltonian_gate.py:125]: Do you want to add `HamiltonianGate` to `qiskit/extensions/__init__.py` so that `QuantumCircuit.hamiltonian` will be added on `import qiskit`? (Right now a user would have to explicitly call `from qiskit.extensions.hamiltonian_gate import HamiltonianGate` or similar before `qc.hamiltonian` will exist.)

File: test/python/circuit/test_hamiltonian_gate.py (hunk 27 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py:136]: I tried to copy the Unitary tests as closely as possible, but I get your point, I've tried to remove some of the clutter.

File: qiskit/extensions/hamiltonian_gate.py (hunk 29 above)
Comment (dongreenberg) [reply qiskit/extensions/hamiltonian_gate.py:125]: Done!

File: test/python/circuit/test_hamiltonian_gate.py (hunk 22 above)
Comment (dongreenberg) [reply test/python/circuit/test_hamiltonian_gate.py]: Removed!

**Comment by dongreenberg** [comment]
@mtreinish would you mind approving? It says I still need your approval as a codeowner for some reason (even though kdk already approved).

**Comment by jaygambetta** [comment]
I am not a big fan of the name as it makes me think it is not simply a unitary with a  Hermitian matrix as the input.  This being said i have wasted 20 min thinking of a name and i cant get a better one so im good.

**Comment by dongreenberg** [comment]
Thanks! Will correct the typo in the changelog
